ir/ir.def

```diff
@@ -500,6 +510,40 @@ class SwitchStatement : Statement {
         split.run_visit(); }
 }
 
+class ForStatement : Statement, ISimpleNamespace {
+    inline IndexedVector<StatOrDecl>    init;
+    Expression                          condition;
+    inline IndexedVector<StatOrDecl>    updates;
+    Statement                           body;
+    IDeclaration getDeclByName(cstring name) const override {
+        return init.getDeclaration(name); }
+    Util::Enumerator<IDeclaration>* getDeclarations() const override {
+        return init.getDeclarations(); }
+    visit_children {
+        v.visit(init, "init");
+        v.visit(condition, "condition");
+        // FIXME -- need visit closure over the body and updates + condition
+        v.visit(body, "body");
+        v.visit(updates, "updates"); }
+}
+
+// FIXME -- should we try to directly convert this to a ForStatement in the parser?
+class ForInStatement : Statement, ISimpleNamespace {
+    NullOK Declaration_Variable         decl;
+    // the declaration will be moved to the enclosing block by MoveDeclarations, but
```

Comment:
I do not recall all the details, but iirc the change in MoveDeclarations was necessary for nested loops like + some shadowing:
```
action a0(bit<8> x, bit<8> y) {
        // Reduce
        bit<64> idx = -1;
        foreach (bit<8> i in 0 .. x) {
            foreach (bit<8> j in i .. y) {
                idx = foo(j);
                if (idx == -1)
                    break;
            }
        }
    }
```